### PR TITLE
Remove recursion in part of Find-References algorithm

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://devdiv.visualstudio.com/DevDiv/_queries/edit/1894307

### CODE REVIEW

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferenceCache.cs [hunk 1]
```diff
@@ -6,195 +6,206 @@
...
-            {
-                // Certain symbols don't have a name, so we return without further searching since the text-based index
-                // and lookup never terminates if searching for an empty string.
-                // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1655431
-                // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1744118
-                // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1820930
-                return ImmutableArray<SyntaxToken>.Empty;
-            }
+        if (_identifierCache.TryGetValue(identifier, out var result))
+            return result;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferenceCache.cs:82]: moved up fast, non-allocating path.

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferenceCache.cs [hunk 2]
```diff
@@ -6,195 +6,206 @@
...
-            // any unicode escapes in it, then we do simple string matching to find the tokens.
-            var info = await SyntaxTreeIndex.GetRequiredIndexAsync(document, cancellationToken).ConfigureAwait(false);
+        // If this document doesn't even contain this identifier (escaped or non-escaped) we don't have to search it at all.
+        if (!info.ProbablyContainsIdentifier(identifier))
+            return ImmutableArray<SyntaxToken>.Empty;
 
-            // If this document doesn't even contain this identifier (escaped or non-escaped) we don't have to search it at all.
-            if (!info.ProbablyContainsIdentifier(identifier))
-                return ImmutableArray<SyntaxToken>.Empty;
+        return await ComputeAndCacheTokensAsync(this, document, identifier, info, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferenceCache.cs:92]: moved closure allocating path to local function.

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferenceCache.cs [hunk 3]
```diff
@@ -6,195 +6,206 @@
...
-            ImmutableArray<SyntaxToken> FindMatchingIdentifierTokensFromTree()
-            {
-                using var _ = ArrayBuilder<SyntaxToken>.GetInstance(out var result);
-                Recurse(root);
-                return result.ToImmutable();
+        static ImmutableArray<SyntaxToken> FindMatchingIdentifierTokensFromTree(
+            ISyntaxFactsService syntaxFacts, string identifier, SyntaxNode root)
+        {
+            using var _ = ArrayBuilder<SyntaxToken>.GetInstance(out var result);
+            using var obj = SharedPools.Default<Stack<SyntaxNodeOrToken>>().GetPooledObject();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferenceCache.cs:124]: meat of hte change is here, moving to a stack approach from a recursion approach.

**Comment by CyrusNajmabadi** [comment]
@akhera99 ptal.

**Review by dibarbet — APPROVED** [review]

**Pull request merged**